pkg/controller/utils/utils.go

```diff
@@ -471,3 +423,59 @@ func StrToElasticLicenseType(license string, logger logr.Logger) render.Elastics
 	logger.V(3).Info("Elasticsearch license %s is unexpected", license)
 	return render.ElasticsearchLicenseTypeUnknown
 }
+
+// WaitToAddResourceWatch will check if projectcalico.org APIs are available and if so, it will add a watch for resource
+// The completion of this operation will be signaled on a ready channel
+func WaitToAddResourceWatch(controller controller.Controller, client kubernetes.Interface, log logr.Logger, flag *ReadyFlag, resourceKind string) {
+	maxDuration := 30 * time.Second
+	duration := 1 * time.Second
+	ticker := time.NewTicker(duration)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			duration = duration * 2
+			if duration >= maxDuration {
+				duration = maxDuration
+			}
+			ticker.Reset(duration)
+			if isResourceReady(client, resourceKind) {
+				err := addResourceWatch(controller, resourceKind)
+				if err != nil {
+					log.Info("failed to watch %s resource: %v. Will retry to add watch", resourceKind, err)
+				} else {
+					flag.MarkAsReady()
+					return
+				}
+			}
+		}
+	}
+}
+
+func addResourceWatch(c controller.Controller, resourceKind string) error {
```

Comment:
I think that instead of using the resource kind and having to switch here, just have `WaitToAddResourceWatch` to accept the object you want to watch. 

Anytime somebody wants to watch a new resource using this they'll have to modify the switch. You've made this slightly more generic in the sense that you can call the same function to watch only two  different resource types, but I think this might be worse then have specific `WaitToAddLicenseKeyWatch` and `WaitToAddDPIKeyWatch` because the developer would have to dig into the internals to figure out that you can only watch these two resources with this.